### PR TITLE
Post acquisition data to support-frontend's backend

### PR DESCRIPTION
**Pull request by svillafe**
## Why are you doing this?

We want to track conversion with the acquisition event.

[**Trello Card**](https://trello.com)

## Changes
* Renamed `acquisition` type to match acquisition team's type. 
* created OphanIds type.
* Added all the data fields to the regular contributions endpoint.

### CODE REVIEW

File: assets/pages/oneoff-contributions/helpers/ajax.js [hunk 1]
```diff
@@ -41,20 +44,27 @@ type OneoffContribFields = {
...
+  const user = state.page.user;
+  const ophanIds: OphanIds = getOphanIds();
+
+  if (user.fullName !== null && user.fullName !== undefined &&
+    user.email !== null && user.email !== undefined) {
+    const referrerAcquisitionData = state.common.referrerAcquisitionData;
 
-  if (state.page.user.fullName !== null && state.page.user.fullName !== undefined
-    && state.page.user.email !== null && state.page.user.email !== undefined) {
     const oneOffContribFields: OneoffContribFields = {
```
Comment (svillafe) [inline assets/pages/oneoff-contributions/helpers/ajax.js]: @Mullefa @desbo We are sending this data to contributions' one-off backend. Is something needed?

File: test/codecs/CirceDecodersTest.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+      referrerAcquisitionData.campaignCode.get mustBe "example"
+      referrerAcquisitionData.abTest.get.name mustBe "test_1"
+      referrerAcquisitionData.abTest.get.variant mustBe "variant_34"
+      referrerAcquisitionData.abTest.get.complete mustBe None
+      referrerAcquisitionData.source.get mustBe GuardianWeb
+      referrerAcquisitionData.componentId mustBe Some("example_component_id")
+      referrerAcquisitionData.componentType mustBe Some(AcquisitionsEpic)
+      referrerAcquisitionData.referrerPageviewId mustBe Some("j8jza3155r9ye3wmfll1")
+      referrerAcquisitionData.componentType
```
Comment (Ap0c) [inline test/codecs/CirceDecodersTest.scala]: Duplicate of line 42?

File: build.sbt [hunk 3]
```diff
@@ -45,8 +45,9 @@ libraryDependencies ++= Seq(
   "com.amazonaws" % "aws-java-sdk-stepfunctions" % "1.11.128",
   "org.typelevel" %% "cats" % "0.9.0",
   "play-circe" %% "play-circe" % "2.6-0.8.0",
-  "com.gu" %% "support-models" % "0.14",
+  "com.gu" %% "support-models" % "0.15-SNAPSHOT",
```
Comment (svillafe) [inline build.sbt]: This should be 0.15. We are waiting for the release of `support-models` to update it.

File: assets/helpers/tracking/acquisitions.js [hunk 4]
```diff
@@ -2,26 +2,36 @@
 
 // ----- Imports ----- //
 
+
+import * as ophan from 'ophan';
```
Comment (Ap0c) [inline assets/helpers/tracking/acquisitions.js]: What happens if ophan gets imported twice on one page? Doesn't it do a bunch of setup stuff on import? Will it do this twice?

File: assets/helpers/tracking/acquisitions.js [hunk 5]
```diff
@@ -2,26 +2,36 @@
...
+
+export type ReferrerAcquisitionData = {|
   campaignCode: ?string,
+  referrerPageviewId: ?string,
   referrerUrl: ?string,
   componentId: ?string,
   componentType: ?string,
   source: ?string,
-  abTests: AcquisitionABTest[],
+  abTest: AcquisitionABTest[],
```
Comment (Ap0c) [inline assets/helpers/tracking/acquisitions.js]: If this is just one AB test, is it still an array?

File: assets/pages/regular-contributions/helpers/ajax.js [hunk 6]
```diff
@@ -20,25 +23,50 @@ const MAX_POLLS = 10;
...
+  const user = inputState.user;
+
+  return user.firstName !== null && user.firstName !== undefined &&
+         user.lastName !== null && user.lastName !== undefined &&
+         user.email !== null && user.email !== undefined;
+};
+
+function removeNullFields(obj: Object): Object {
+
+  const response = obj;
```
Comment (Ap0c) [inline assets/pages/regular-contributions/helpers/ajax.js]: Isn't this copy by reference? You're modifying the original object here as well, no?

File: app/codecs/CirceDecoders.scala [hunk 7]
```diff
@@ -58,9 +62,24 @@ object CirceDecoders {
     Decoder.decodeString.emap(code => BillingPeriod.fromString(code).toRight(s"Unrecognised period code '$code'"))
   implicit val encodePeriod: Encoder[BillingPeriod] = Encoder.encodeString.contramap[BillingPeriod](_.toString)
 
+  implicit val ophanIdsCodec: Codec[OphanIds] = deriveCodec
```
Comment (Mullefa) [inline app/codecs/CirceDecoders.scala]: You should be able to import most of these implicits from acquisitions event producer.

File: assets/helpers/tracking/acquisitions.js (hunk 4 above)
Comment (svillafe) [reply assets/helpers/tracking/acquisitions.js]: It only sends one pageview. 
picture 412

File: app/codecs/CirceDecoders.scala (hunk 7 above)
Comment (Mullefa) [reply app/codecs/CirceDecoders.scala]: Ok, as discussed, run as is and refactor in a separate pull request.

File: test/codecs/CirceDecodersTest.scala [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        """
+          .stripMargin
+
+      val parsedJson2 = parse(json2).toOption.get
+
+      val referrerAcquisitionData2: ReferrerAcquisitionData = referrerAcquisitionDataCodec.decodeJson(parsedJson2).right.get
+
+      referrerAcquisitionData2.componentType mustBe Some(EnumUnknownComponentType(-1))
+
+      referrerAcquisitionData2.componentType
```
Comment (Ap0c) [inline test/codecs/CirceDecodersTest.scala]: Duplicate of line 57?

File: assets/pages/oneoff-contributions/helpers/ajax.js (hunk 1 above)
Comment (Mullefa) [reply assets/pages/oneoff-contributions/helpers/ajax.js]: This is ok, but we include referrer acquisition data in an addition request? See [`ContributionRequest`](https://github.com/guardian/contributions-frontend/blob/master/app/controllers/forms/ContributionRequest.scala) for all fields currently accepted.

**Comment by svillafe** [comment]
Tested in Code using https://github.com/guardian/support-workers/pull/60 workers in Code.

**Comment by svillafe** [comment]
Acceptance tests running locally 🎉

**Comment by svillafe** [comment]
Tested in Prod (due to a domain constraint) we are successfully retrieving the browser id and the visit it.

picture 414

**Pull request merged**